lessons/12-redux/README.md

```diff
@@ -1 +1,83 @@
+## History
+While Facebook was making React, they also came out with [Flux](https://github.com/facebook/flux).  Flux is a pattern for managing state and data flow in React applications.  It is not a strict framework, but more of an idea for a pattern that Facebook think developers should use.
+
+Since Flux is not its own framework, many different frameworks have come out to try and implement the Flux pattern.  Redux is not a strict implementation of Flux ([depending on who you ask](http://redux.js.org/docs/introduction/PriorArt.html#flux)), but it is heavily inspired by the Flux architecture.  We'll be using Redux for our apps in this class.  Unless you're curious, there's no need to learn Flux itself.
+
+## Overview
+### Three Core Principles
+1) The whole state of your app is contained in a single state object in a **store**.
+2) When the user does something that should change the state, an **action** is dispatched.
+3) The action is sent through a **reducer** function, which specifies how to change the state.
+
+### Todo Example
+If you're making a Todo app, the state of your app might look like this:
+```javascript
+{
+  filter: 'COMPLETED',
+  todos: [
+    {
+      text: 'brush teeth',
+      completed: true,
+    },
+    {
+      text: 'do my JS homework',
+      completed: false
+    }
+  ]
+}
+```
+We have two todos, one completed and one not, and the filter is set so that the user only sees the "completed" todos.
+
+Say the user changes the filter by pressing the "active" button.  The React/Redux app then dispatches an **action** that looks like this:
+```javascript
+var actionObj = {
+  type: 'CHANGE_FILTER',
+  filter: 'ACTIVE'
+};
+```
+This is the **action object**.  It says that a `'CHANGE_FILTER'` action was performed and also gives some relevant data (in this case, what the new filter is).
+
+The action object will then be passed into a **dispatcher** to be dispatched:
+```javascript
+dispatch(actionObj);
+```
+
+The **reducer** then receives the current state and the action object as we defined above, and returns the new state:
+```javascript
+var todoReducer = function(state, action) {
+  switch (action.type) {
+    case 'CHANGE_FILTER':
+      return Object.assign({}, state, {
+        filter: action.filter
+      });
+
+    default:
+      return state;
+  }
+}
+```
+Normally a reducer has multiple `case`s in its `switch` statement to handle all the different action types.
+
+The [`Object.assign()` function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) merges multiple object together.  In reducers, this is often used to merge the old state with the new properties that we change. In this case, the new property is that `filter` is now set to `'ACTIVE'`.  This new `{filter: 'ACTIVE'}` object is merged with the old state and overrides the previous filter while still containing all our todos.
```

Comment:
Might mention that you always return a new state object instead of modifying the current one and that Object.assign() and (if they use babel to get object destructuring) {...state} both return new objects. Yay immutability